Improvements to Teleport PAM support.

russjones opened the pull request:
#### Description

Added ability to read in PAM environment variables from PAM handle and pass environment variables to PAM module `TELEPORT_USERNAME`, `TELEPORT_LOGIN`, and `TELEPORT_ROLES`.

Refactored launching of shell to call PAM first. This allows a PAM module to create the user and home directory before attempting to launch a shell for said user.

To do this the command passed to Teleport during re-exec has changed. Before the Teleport master process would resolve the user fully (UID, GUID, supplementary groups, shell, home directory) before re-launching itself to then launch a shell. However, if PAM is used to create the user on the fly and PAM has not been called yet, this will fail.

Instead that work has now been pushed to occur in the child process. This means the Teleport master process now creates a payload with the minimum needed from `*srv.ServerContext` and will then re-exec itself. The child process will call PAM and then attempt to resolve the user (UID, GUID, supplementary groups, shell, home directory).

#### Examples

##### Using `pam_exec.so`

Using `pam_exec.so` is the easiest way to use the PAM stack to create a user if the user does not already exist. The advantage of using `pam_exec.so` is that it usually ships with the operating system. The downside is that it doesn't provide access to some additional environment variables that Teleport sets (see the `pam_script.so` example for those) to use additional identity metadata in the user creation process.

You can either add `pam_exec.so` to your existing PAM stack for your application or write a new one for Teleport. In this example we'll write a new one to simplify how to use `pam_exec.so` with Teleport.

Start by creating a file `/etc/pam.d/teleport` with the following contents.

```
account   required   pam_exec.so /etc/pam-exec.d/teleport_acct
session   required   pam_motd.so
```

Note the inclusion of `pam_motd.so` under the session facility. While `pam_motd.so` is not required for user creation, Teleport requires a module set for both the `account` and `session` facility to work.

Next create the script that will be run by `pam_exec.so` like below. This script will check if the user passed in `PAM_USER` exists and if it does not, it will create it. Any error from `useradd` will be written to `/tmp/pam.error`.

```
mkdir -p /etc/pam-exec.d
cat > /etc/pam-exec.d/teleport_acct <<EOF
#!/bin/sh
id -u "${PAM_USER}" &>/dev/null  || /sbin/useradd -m "${PAM_USER}" 2> /tmp/pam.error
exit 0
EOF
chmod +x /etc/pam-exec.d/teleport_acct
```

Next update `/etc/teleport.yaml` to call the above PAM stack by both enabling PAM and setting the `service_name`.

```
ssh_service:
   pam:
     enabled: yes
     service_name: "teleport"
```

Now attempting to login as an existing user should result in the creation of the user and successful login.

##### Using `pam_script.so`

If more advanced functionality is needed `pam_script.so` is a good choice. It typically has to be installed from packages but richer scripts with more identity information from Teleport can be used during the process of user creation.

To start install `pam_script.so`. On Debian based systems this would be `apt-get install libpam-script` and on RHEL based systems `yum install pam-script`.

You can either add `pam_script.so` to your existing PAM stack for your application or write a new one for Teleport. In this example we'll write a new one to simplify how to use `pam_script.so` with Teleport.

Start by creating a file `/etc/pam.d/teleport` with the following contents.

```
account   required   pam_script.so
session   required   pam_motd.so
```

Note the inclusion of `pam_motd.so` under the session facility. While `pam_motd.so` is not required for user creation, Teleport requires a module set for both the `account` and `session` facility to work.

Next create the script that will be run by `pam_script.so` like below. This script will check if the user passed in `TELEPORT_LOGIN` exists and if it does not, it will create it. Any error from `useradd` will be written to `/tmp/pam.error`. Note the additional environment variables `TELEPORT_USERNAME`, `TELEPORT_ROLES`, and `TELEPORT_LOGIN`. These can be used to write richer scripts that may change the system in other ways based off identity information.

```
mkdir -p /etc/pam-script.d
cat > /etc/pam-script.d/teleport_acct <<EOF
#!/bin/sh
COMMENT="User ${TELEPORT_USERNAME} with roles ${TELEPORT_ROLES} created by Teleport."
id -u "${TELEPORT_LOGIN}" &>/dev/null  || /sbin/useradd -m -c "${COMMENT}" "${TELEPORT_LOGIN}" 2> /tmp/pam.error
exit 0
EOF
chmod +x /etc/pam-script.d/teleport_acct
```

Next update `/etc/teleport.yaml` to call the above PAM stack by both enabling PAM and setting the `service_name`.

```
ssh_service:
   pam:
     enabled: yes
     service_name: "teleport"
```

Now attempting to login as an existing user should result in the creation of the user and successful login.

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/3270
Fixes https://github.com/gravitational/teleport/issues/3021
Fixes https://github.com/gravitational/teleport/issues/2863

russjones committed:
Added support for PAM environment variables.

Added method that reads in PAM environment variables from PAM handle.

russjones committed:
Added ability to pass metadata to PAM modules.

Added ability to pass additional metadata about the user to PAM modules
through the PAM_RUSER field.

russjones committed:
Added more test support to pam_teleport.so.

Added support for "pam_putenv" and "pam_get_item" to fetch PAM_RUSER to
pam_teleport.so. This is used for test coverage.

russjones committed:
Added testing.Verbose to allow silencing of tests.

russjones committed:
Refactored launching of shell.

Refactored launching of shell to call PAM first. This allows a PAM
module to create the user and home directory before attempting to launch
a shell for said user.

To do this the command passed to Teleport during re-exec has changed.
Before the Teleport master process would resolve the user fully (UID,
GUID, supplementary groups, shell, home directory) before re-launching
itself to then launch a shell. However, if PAM is used to create the
user on the fly and PAM has not been called yet, this will fail.

Instead that work has now been pushed to occur in the child process.
This means the Teleport master process now creates a payload with the
minimum needed from *srv.ServerContext and will then re-exec itself. The
child process will call PAM and then attempt to resolve the user (UID,
GUID, supplementary groups, shell, home directory).

webvictim commented on lib/pam/pam.go:
```diff
@@ -35,7 +35,11 @@ package pam
 // extern int _pam_acct_mgmt(void *, pam_handle_t *, int);
 // extern int _pam_open_session(void *, pam_handle_t *, int);
 // extern int _pam_close_session(void *, pam_handle_t *, int);
+// extern char **_pam_getenvlist(void *handle, pam_handle_t *pamh);
```
No spacing between `**` and the function name - presumably this is all flattened out by the compiler anyway because function names can't contain `**`?

russjones replied:
We don't really have a C style guide that we follow as far as I know. That being said, but I have been trying to follow how it's declared in the PAM man pages.

```
NAME
       pam_getenvlist - getting the PAM environment

SYNOPSIS
       #include <security/pam_appl.h>

       char **pam_getenvlist(pam_handle_t *pamh);
```

webvictim commented on lib/pam/pam.c:
```diff
@@ -207,6 +207,30 @@ int _pam_close_session(void *handle, pam_handle_t *pamh, int flags)
     return (f)(pamh, flags);
 }
 
+char **_pam_getenvlist(void *handle, pam_handle_t *pamh)
```
See comment in pam.go

webvictim commented on lib/pam/pam_test.go:
```diff
@@ -0,0 +1,125 @@
+// echo the contents of PAM_RUSER to stdout where this test can read, parse,
+// and validate it's output.
+func (s *Suite) TestEcho(c *check.C) {
+	// Skip this test if the binary was not build with PAM support.
```
`not build with` -> `not built with`

webvictim replied:
`was not build` -> `was not built`

webvictim commented on lib/pam/pam_test.go:
```diff
@@ -0,0 +1,125 @@
+//
+// The PAM module used, pam_teleport.so is called from the policy file
+// teleport-session-environment. The policy file instructs pam_teleport.so to
+// read in the first argument and set it as a PAM environment variaable. This
```
`variaable` -> `variable`

webvictim replied:
`variaable` -> `variable`

webvictim commented on lib/pam/pam_test.go:
```diff
@@ -0,0 +1,125 @@
+// read in the first argument and set it as a PAM environment variaable. This
+// test then validates it matches what was set in the policy file.
+func (s *Suite) TestEnvironment(c *check.C) {
+	// Skip this test if the binary was not build with PAM support.
```
`not build with` -> `not built with`

webvictim replied:
`was not build` -> `was not built`

webvictim commented on lib/srv/ctx.go:
```diff
@@ -630,6 +630,92 @@ func (c *ServerContext) String() string {
+		roleNames = c.Identity.RoleSet.RoleNames()
+	}
+
+	// Create the execCommand that will be send to the child process.
```
`send` -> `sent`

webvictim replied:
`will be send` -> `will be sent`

webvictim commented on lib/srv/ctx.go:
```diff
@@ -630,6 +630,92 @@ func (c *ServerContext) String() string {
+	// Apply environment variables passed in from client.
+	for k, v := range ctx.env {
+		env = append(env, fmt.Sprintf("%s=%s", k, v))
+	}
```
Will this also contain anything present in `/etc/environment`? Does that get read in earlier?

russjones replied:
These are environment variables passed by the client. They are saved in the context then applied here.

https://github.com/gravitational/teleport/blob/master/lib/srv/regular/sshserver.go#L1126-L1127
https://github.com/gravitational/teleport/blob/master/lib/srv/regular/sshserver.go#L1235-L1245

The PAM environment variables are applied during the re-exec, you can see that here:

https://github.com/gravitational/teleport/blob/rjones/pam-cgo/lib/srv/exec.go#L367-L368
https://github.com/gravitational/teleport/blob/rjones/pam-cgo/lib/srv/exec.go#L567-L568

Your comment also raised another interesting question, if a duplicate environment variables exist, which one should win? The one that comes from the client or the one from PAM? I checked what OpenSSH does and the one from PAM wins.

In our case since we apply PAM environment variables last, [they should win](https://golang.org/pkg/os/exec/#Cmd):

```
    // If Env contains duplicate environment keys, only the last
    // value in the slice for each duplicate key is used.
    // As a special case on Windows, SYSTEMROOT is always added if
    // missing and not explicitly set to the empty string.
```

webvictim commented on lib/srv/ctx.go:
```diff
@@ -630,6 +630,92 @@ func (c *ServerContext) String() string {
+	// Apply environment variables passed in from client.
+	for k, v := range ctx.env {
+		env = append(env, fmt.Sprintf("%s=%s", k, v))
+	}
```
Will this handle/contain anything listed in `/etc/environment` too? I can't see where we'd be reading that in.

russjones replied:
Answered here https://github.com/gravitational/teleport/pull/3317#discussion_r373207558

webvictim commented on lib/srv/ctx.go:
```diff
@@ -630,6 +630,92 @@ func (c *ServerContext) String() string {
+		}
+	}
+
+	// Set some Teleport specific environment variable: SSH_TELEPORT_USER,
```
`variable` -> `variables`

webvictim commented on modules/pam_teleport/pam_teleport.c:
```diff
@@ -22,6 +24,31 @@ int pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc, const char **argv)
 {
+    int pam_err;
+
+    // If the "echo" command is requested that will echo out the value of
```
"echo" -> "echo_ruser"

webvictim reviewed (commented):
LGTM in theory with a few nits

fspmarshall commented on lib/pam/pam.go:
```diff
@@ -311,11 +338,45 @@ func (p *PAM) Close() error {
 	C.free(unsafe.Pointer(p.service_name))
-	C.free(unsafe.Pointer(p.user))
+	C.free(unsafe.Pointer(p.login))
+	C.free(unsafe.Pointer(p.loginContext))
```
If `Close` returns an error, these `free` calls never run and (presumably) the values leak.  Is it possible to `defer` these at the beginning of `Close`, or is it unsafe to `free` these unless all of the pam close/end stuff was successful?

russjones commented:
@gravitational-jenkins retest this please

fspmarshall reviewed (commented):
`pam_teleport.so` is being tracked.  This is probably unintended.

russjones committed:
Fix memory leak and pass variables to PAM.

russjones committed:
Don't call defer from function that exits.

Don't call os.Exit() from RunCommand() as any defers won't be called.
Instead wrap RunCommand() in RunAndExit() to allow defers to be called.

russjones commented:
@gravitational-jenkins retest this please

russjones commented:
@gravitational-jenkins retest this please
